vm/src/sliceable.rs

```diff
@@ -291,50 +292,49 @@ impl SequenceIndex {
     }
 }
 
-// Use PySliceableSequence::wrap_index for implementors
-pub fn wrap_index(p: isize, len: usize) -> Option<usize> {
-    let neg = p.is_negative();
-    let p = p.wrapping_abs() as usize;
-    if neg {
-        len.checked_sub(p)
-    } else if p >= len {
-        None
-    } else {
-        Some(p)
-    }
+pub trait SequenceIndexOp {
+    // Saturate p in range [0, len] inclusive
+    fn saturated_at(&self, len: usize) -> usize;
+    // Use PySliceableSequence::wrap_index for implementors
+    fn wrapped_at(&self, len: usize) -> Option<usize>;
 }
 
-// Saturate p in range [0, len] inclusive
-pub fn saturate_index(p: isize, len: usize) -> usize {
-    let len = len.to_isize().unwrap_or(isize::MAX);
-    let mut p = p;
-    if p < 0 {
-        p += len;
-        if p < 0 {
-            p = 0;
-        }
+impl SequenceIndexOp for isize {
+    fn saturated_at(&self, len: usize) -> usize {
+        let len = len.to_isize().unwrap_or(Self::MAX);
+        *self.clamp(&0, &len) as usize
```

Comment:
I didn't mean to remove `if p < 0` completely.

```python
let mut p = *self;
if p < 0 {
    p += len;
}
```

Something like :point_up: should be done before clamp() to handle `p <  0 and 0 <= p + len < len` case.